R/sheets_speedread.R

```diff
@@ -0,0 +1,221 @@
+#'
+#' # read via 'datasource' URL (access token goes as query param)
+#' sheets_speedread(ss, .url = "datasource")
+#'
+#' # clean up
+#' googledrive::drive_trash(ss)
+#' }
+sheets_speedread <- function(ss,
+                             sheet = NULL,
+                             range = NULL,
+                             skip = 0,
+                             ...,
+                             .url = c("export", "datasource"),
+                             .method = c("vroom", "readr")) {
+  .method <- match.arg(.method)
+  stop_for_pkg <- function(pkg) {
+    if (!requireNamespace(pkg, quietly = TRUE)) {
+      stop_glue("The {pkg} package must be installed to use {bt('sheets_speedread()'}")
+    }
+  }
+  stop_for_pkg(.method)
+
+  ssid <- as_sheets_id(ss)
+  maybe_sheet(sheet)
+  check_range(range)
+  check_non_negative_integer(skip)
+  .url <- match.arg(.url)
+
+   x <- sheets_get(ssid)
+
+  params <- list(spreadsheet_id = unclass(ssid))
+  sheet_msg <- ""
+  range_msg <- ""
+  range_spec <- as_range_spec(
+    range, sheet = sheet, skip = skip,
+    sheets_df = x$sheets, nr_df = x$named_ranges
+  )
+  if (!is.null(range_spec$named_range)) {
+    stop_glue("{bt('sheets_speedread()'} cannot work with a named range")
+  }
+  if (!is.null(range_spec$cell_limits)) {
+    range_spec$cell_range <- as_sheets_range(range_spec$cell_limits)
+  }
+  if (!is.null(range_spec$cell_range)) {
+    params <- c(params, range = range_spec$cell_range)
+    range_msg <- glue(", range {dq(range_spec$cell_range)}")
+  }
+  if (!is.null(range_spec$sheet_name)) {
+    s <- lookup_sheet(range_spec$sheet_name, sheets_df = x$sheets)
+    params <- c(params, gid = s$id)
+    sheet_msg <- glue(", {dq(range_spec$sheet_name)} sheet")
+  }
+  message_glue("Reading from {dq(x$name)}{sheet_msg}{range_msg}")
+
+  token <- sheets_token()
+
+  if (.url == "export") {
+    params[["format"]] <- "csv"
+    params[["req_path"]] <- "export"
+  }
+  if (.url == "datasource") {
+    params[["tqx"]] <- "out:csv"
+    params[["req_path"]] <- "gviz/tq"
+    if (!is.null(token)) {
+      token$auth_token$refresh()
+      params[["access_token"]] <- token$auth_token$credentials$access_token
+    }
+  }
+
+  req <- gargle::request_build(
+    path = "spreadsheets/d/{spreadsheet_id}/{req_path}",
+    method = "GET",
+    params = params,
+    base_url = "https://docs.google.com"
+  )
+
+  if (.url == "datasource" || is.null(token)) {
+    if (.method == "vroom") {
+      return(vroom::vroom(req$url, delim = ",", ...))
+    } else { # .method == "readr"
+      return(readr::read_csv(req$url, ...))
+    }
+  }
+
+  # .url == "export" and we must have a token
+  response <- httr::GET(req$url, config = token)
+  stopifnot(identical(httr::http_type(response), "text/csv"))
+  csv <- httr::content(response, as = "text", encoding = "UTF-8")
+  if (.method == "vroom") {
+    vroom::vroom(csv, delim = ",", ...)
+  } else {
+    readr::read_csv(csv, ...)
+  }
```

Comment:
I feel like I should be able to do this with `httr::with_config()` (for the token), but I just couldn't get it to work.